### ISSUE TITLE
Use preview namespace for regular workers

### ISSUE THREAD

**Issue by EverlastingBugstopper** [cloudflare/wrangler#1032]
For Workers Sites, we create a separate preview namespace so that we do not impact production namespaces. We do not do this for regular workers and we should.

**Comment by GregBrimble** [comment]
The option to duplicate data into a new namespace would be great! It would be really useful to let developers spawn a development environment from production without needing to [@cloudflare/kv-worker-migrate](https://github.com/cloudflare/kv-worker-migrate).

**Comment by ashleymichal** [comment]
> 1. Start creating a namespace automatically on publish if it doesn't exist? We'll have to start doing this on preview and it seems... strange to have different behavior on preview/publish. If we're smart enough on preview, why not publish?
the reason we can do this for sites is that we control the title of the namespace; one mistake i think we made with KV was to use ID instead of Title in the wrangler.toml; Titles are unique per account and therefore would have been better. This particular point would work great if we remedied that by taking a `title` argument instead of an `id` argument in the toml, making kv-namespaces top level, and concatenating environment names for envs the way we do for site namespaces. if not, all we have is binding name, and this should not be coupled to title, as the user should be allowed to change it without unexpected effects on their namespace(s), tho perhaps the same could be said for envs.
> 2. Add a --preview flag to all of the KV namespace commands?
probably couldn't hurt? tangent: what if `preview` was a reserved environment name? 🤔 
> 3. Seed the preview namespace from production data on startup (copy over all the data)?
this is complex and difficult for a few reasons. providing tools for users to do this i think could be valuable but doing so automatically might be overkill.
> 4. Create a new namespace automatically on preview sessions based on namespaces defined in wrangler.toml?
this seems good, given the same constraints as my response to point 1
> Delete old values in the namespace when the process is killed?
depending on our concerns around churn, we could simply delete preview namespaces when the process is killed; this would be simplest but slowest. making the options around this configurable might be valuable.

**Comment by EverlastingBugstopper** [comment]
> This particular point would work great if we remedied that by taking a title argument instead of an id argument in the toml, making kv-namespaces top level, and concatenating environment names for envs the way we do for site namespaces.

unfortunately this has some major backwards incompatibility issues - today we take id and binding and don't allow inheritance from top level. 

> probably couldn't hurt?

probably would be reasonable addition but would need some way to know if preview session currently exists

> what if `preview` was a reserved environment name

this is an interesting thought, and one that we had discussed in the original environments proposal - dont think it's a good idea here though given backwards compatibility issues and also i'd assume you'd want to use wrangler dev across multiple environments

> this is complex and difficult for a few reasons. providing tools for users to do this i think could be valuable but doing so automatically might be overkill.

i think you're right and probably not necessary for mvp

> depending on our concerns around churn, we could simply delete preview namespaces when the process is killed; this would be simplest but slowest. making the options around this configurable might be valuable.

probably makes sense just to delete entirely.

---

The biggest concern I have here is backwards compatibility - changing the way we configure kv namespaces in the pursuit of wrangler dev feels... icky. Perhaps what we can do is

1) fetch title of kv namespace from id in toml
2) prepend `__preview_` to title and create new namespace
3) use that namespace w/same binding for preview worker
4) delete preview namespace when session ends

I think `--preview` flag on kv commands is probably too complicated and error-prone to pursue, would likely want to print out the title of the namespace we're using and say something like 

`You can run kv commands using the --title __preview_your_namespace_id flag`

**Comment by ashleymichal** [comment]
> unfortunately this has some major backwards incompatibility issues - today we take id and binding and don't allow inheritance from top level.

agreed; this might be a thing to stuff into a 2.0 milestone, which we may want to consider compiling.

the sequence you have is pretty good; it does rely on a clean up step and that really leaves room for odd state problems (what do you do if the preview namespace still exists?)

i think it is worth exploring the use of kv prefixing for this...

**Comment by EverlastingBugstopper** [comment]
ok so - when a preview session is started with an environment that contains kv-namespaces, we will 

1) fetch the title of the kv namespace from the id in the toml
2) prepend the title with `__preview_` and create the new namespace if it doesn't exist
3) print the name of the namespace we're using and make sure they know it is separate from production and they may need to repopulate it
4) when uploading the worker to the preview api, use the same binding name for the namespace and new namespace id
5) on teardown, inform the user that their preview namespace still exists and they can delete it if they want to

**Comment by EverlastingBugstopper** [comment]
I've updated the [top level comment](https://github.com/cloudflare/wrangler/issues/1032#issue-560622548) to reflect the current state of the discussion.

**Comment by ispivey** [comment]
Good feedback/question from @koeninger: "What do I do if I want to preview my production KV namespace?"

**Comment by ashleymichal** [comment]
could we add an optional `preview-id` field to each kv namespace entry

**Comment by EverlastingBugstopper** [comment]
ok - i've kinda dropped the ball on actually starting implementation on this one/nailing down exact design.

@ashleymichal I really like the idea of being explicit here and adding the idea of a preview namespace to `wrangler.toml`. i also think we could combine it with environment variables to allow individual developers to each have their own preview namespace, which would solve the problem of having multiple developers squashing each others namespaces.

i think to start out with, we should just flat out refuse to preview if there is not a preview namespace specified for the environment they're using and make them handle the creation/deletion themselves. this makes it so we are not in charge of provisioning namespaces or causing a bunch of churn/inflating their KV billing.

**Comment by ashleymichal** [comment]
@EverlastingBugstopper this seems like a good call. we can keep our eyes  out for optimizations  and convenience features, but for solving the biggest problem this seems good.
